components/style/values/specified/position.rs

```diff
@@ -8,616 +8,247 @@
-                    PositionCategory::HorizontalKeyword |
-                    PositionCategory::HorizontalLogicalKeyword => return Err(()),
-                    _ => Some(key),
-                }
-            },
-            Some(_) => return Err(()),
-            None => None,
-        };
-
-        // Unwrapping and checking position.
-        let position = match position {
-            Some(PositionComponent::Length(pos)) => {
-                // "center <length>" is not allowed
-                if let Some(Keyword::Center) = keyword {
-                    return Err(());
-                }
-                Some(pos)
-            },
-            Some(_) => return Err(()),
-            None => None,
-        };
-
-        Ok(VerticalPosition {
-            keyword: keyword,
-            position: position,
+    fn from_computed_value(computed: &computed_position::HorizontalPosition) -> HorizontalPosition {
+        GenericHorizontalPosition(PositionValue {
+            keyword: None,
+            position: Some(ToComputedValue::from_computed_value(&computed.0)),
         })
     }
 }
 
+/// The specified value of vertical `<position>`
+pub type VerticalPosition = GenericVerticalPosition<PositionValue<LengthOrPercentage>>;
+
 impl ToComputedValue for VerticalPosition {
     type ComputedValue = computed_position::VerticalPosition;
 
     #[inline]
     fn to_computed_value(&self, context: &Context) -> computed_position::VerticalPosition {
-        let keyword = self.keyword.unwrap_or(Keyword::Left);
-
-        // Construct vertical computed LengthOrPercentage
-        let vertical = match keyword {
-            // FIXME(canaltinova): Support logical keywords.
-            Keyword::Bottom | Keyword::YEnd => {
-                if let Some(ref x) = self.position {
-                    let (length, percentage) = match *x {
-                        LengthOrPercentage::Percentage(Percentage(y)) => (Au(0), Some(1.0 - y)),
-                        LengthOrPercentage::Length(ref y) => (-y.to_computed_value(context), Some(1.0)),
-                        _ => (Au(0), None),
-                    };
-                    ComputedLengthOrPercentage::Calc(CalcLengthOrPercentage {
-                        length: length,
-                        percentage: percentage
-                    })
-                } else {
-                    ComputedLengthOrPercentage::Percentage(1.0)
-                }
-            },
-            Keyword::Center => {
-                keyword.to_length_or_percentage().to_computed_value(context)
-            },
-             _ => {
-                let zero = LengthOrPercentage::Percentage(Percentage(0.0));
-                let vert = self.position.as_ref().unwrap_or(&zero);
-                vert.to_computed_value(context)
-            },
-        };
-
-        computed_position::VerticalPosition(vertical)
+        GenericVerticalPosition(self.0.computed_value(context, |keyword| {
+            match keyword {
```

Comment:
Yeah, it's wrong, it should've been different. Good catch! But, I realize that we don't really need a function at all. Since we already parse keywords properly, I've added a function to `Keyword` to check for the opposite sides.